Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt

```diff
@@ -81,6 +85,59 @@ class HomeFragment : Fragment() {
     dialog.show()
   }
 
+  private fun setupSyncUploadProgress() {
+    lifecycleScope.launch {
+      mainActivityViewModel.syncUploadState.collect {
+        when (it) {
+          is SyncUploadState.Started,
+          is SyncUploadState.InProgress -> showSyncBanner(it)
+          is SyncUploadState.Completed -> hideSyncBanner(it)
+          is SyncUploadState.Failed -> hideSyncBanner(it)
+          is SyncUploadState.NoOp -> {
+            println("No operation required. Sync is in progress")
+          }
+        }
+      }
+    }
+  }
+
+  private fun showSyncBanner(syncUploadState: SyncUploadState) {
+    with(binding.uploadLayout) {
+      if (linearLayoutUploadStatus.visibility != View.VISIBLE) {
+        // may add fade in animation here later
+        linearLayoutUploadStatus.visibility = View.VISIBLE
+      } else if (syncUploadState is SyncUploadState.InProgress) {
+        updateUploadPercent(syncUploadState.completedRequests, syncUploadState.totalRequests)
+      }
+    }
+  }
+
+  private fun hideSyncBanner(syncUploadState: SyncUploadState) {
+    if (syncUploadState is SyncUploadState.Completed) {
+      updateUploadPercent(syncUploadState.totalRequests, syncUploadState.totalRequests)
+      binding.uploadLayout.uploadPercent.text = "Uploaded"
+      binding.uploadLayout.linearLayoutUploadStatus.visibility = View.GONE
+    } else if (syncUploadState is SyncUploadState.Failed) {
+      binding.uploadLayout.uploadPercent.text = "Failed"
+      // we can add some animation here
+      binding.uploadLayout.linearLayoutUploadStatus.visibility = View.GONE
```

Comment:
How about providing an error message as well?

Reply:
We give a toast saying "Upload Failed"